### ISSUE TITLE
Horizontal Attention plot at synthesis

### ISSUE THREAD

**Issue by noetits** [Kyubyong/dc_tts#27]
If you try, in synthesis, to save and show Attention computed with the model pretrained on LJ-speech for example, it will look like this:
![alignment_3](https://user-images.githubusercontent.com/11230320/45104081-85235800-b131-11e8-93c4-bb431a367f5d.png)

Why is it horizontal and not diagonal like during the training ? The synthesis works just fine though ...

If I comment, in "networks.py", in the function "Attention" the part corresponding to "monotonic attention" like this:
```
    A = tf.matmul(Q, K, transpose_b=True) * tf.rsqrt(tf.to_float(hp.d))
    # if mononotic_attention:  # for inference
    #     key_masks = tf.sequence_mask(prev_max_attentions, hp.max_N)
    #     reverse_masks = tf.sequence_mask(hp.max_N - hp.attention_win_size - prev_max_attentions, hp.max_N)[:, ::-1]
    #     masks = tf.logical_or(key_masks, reverse_masks)
    #     masks = tf.tile(tf.expand_dims(masks, 1), [1, hp.max_T, 1])
    #     paddings = tf.ones_like(A) * (-2 ** 32 + 1)  # (B, T/r, N)
    #     A = tf.where(tf.equal(masks, False), A, paddings)
    A = tf.nn.softmax(A) # (B, T/r, N)
    max_attentions = tf.argmax(A, -1)  # (B, T/r)
    R = tf.matmul(A, V)
    R = tf.concat((R, Q), -1)
```

The attention plot will be of diagonal shape, and the synthesis not too bad but will have the problem mentioned in the paper: may skip letters or pronounce several times parts of words.

**Comment by noetits** [comment]
Coming back on this, I finally figured this out. In the monotonic attention, he put a mask around prev_max_attention indices and he set everything outside that mask to a big negative number: -2**32+1

That way, after the softmax, we are sure that the max attention will be around previous_max_attention.

The thing is that, to have the correct attention plot at the end, we should use this trick only on the last frame of A, and not on all frames as done here. This changes nothing for the synthesis and I guess that is the reason why @Kyubyong did not bother.

Anyway, if someone is interested to have the correct Attention plot, this should work:
```
if mononotic_attention:  # for inference
    key_masks = tf.sequence_mask(prev_max_attentions, hp.max_N)
    reverse_masks = tf.sequence_mask(hp.max_N - hp.attention_win_size - prev_max_attentions, hp.max_N)[:, ::-1]
    masks = tf.logical_or(key_masks, reverse_masks)
    # masks = tf.tile(tf.expand_dims(masks, 1), [1, hp.max_T, 1])
    masks = tf.tile(tf.expand_dims(masks, 1), [1, 1, 1])
    Falses=tf.fill((tf.shape(A)[0], tf.shape(A)[1]-1, tf.shape(A)[2]), False)
    masks=tf.concat([Falses, masks], axis=1)
    paddings = tf.ones_like(A) * (-2 ** 32 + 1)  # (B, T/r, N)
    A = tf.where(tf.equal(masks, False), A, paddings)
```

**Comment by inventor617** [comment]
@noetits Hi, how did you solve the problem?What kind of verions of python, GPU did you use? I have the same problem.
